### PR TITLE
Fix ARM32 Stalker's handling of large IT blocks

### PR DESCRIPTION
**Pull request by WorksButNotTested**

### CODE REVIEW

File: gum/arch-arm/gumthumbrelocator.c [hunk 1]
```diff
@@ -863,19 +853,31 @@ gum_thumb_relocator_rewrite_it_block_start (GumThumbRelocator * self,
...
+  {
+    case GUM_THUMBRELOCATOR_BRANCH_SHORT:
+      gum_thumb_writer_put_b_cond_label (self->output, cc, block->then_label);
+      break;
+    case GUM_THUMBRELOCATOR_BRANCH_LONG:
+      gum_thumb_writer_put_b_cond_label_wide (self->output, cc,
+          block->then_label);
+      break;
+    default:
+      g_assert_not_reached();
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:877]:
```suggestion
      g_assert_not_reached ();
```

File: gum/arch-arm/gumthumbrelocator.c [hunk 2]
```diff
@@ -885,26 +887,29 @@ static void
...
+  switch (self->it_branch_type)
+  {
+    case GUM_THUMBRELOCATOR_BRANCH_SHORT:
+      gum_thumb_writer_put_b_label (self->output, block->end_label);
+      break;
+    case GUM_THUMBRELOCATOR_BRANCH_LONG:
+      gum_thumb_writer_put_b_label_wide (self->output, block->end_label);
+      break;
+    default:
+      g_assert_not_reached();
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:900]:
```suggestion
      g_assert_not_reached ();
```

File: gum/arch-arm/gumthumbrelocator.c [hunk 3]
```diff
@@ -885,26 +887,29 @@ static void
 gum_thumb_relocator_rewrite_it_block_else (GumThumbRelocator * self,
                                            GumThumbITBlock * block)
 {
-  gum_thumb_location_init (&block->else_branch,
-      gum_thumb_writer_cur (self->output), self->output->pc);
-
-  gum_thumb_writer_put_instruction (self->output, 0xe000);
+  block->end_label = self->output->code + 1;
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:890]:
```suggestion
  block->end_label = self->output->code + 1;

```

File: gum/arch-arm/gumthumbrelocator.c [hunk 4]
```diff
@@ -969,8 +974,19 @@ gum_parse_it_instruction_block_size (guint16 insn)
...
+gum_thumb_relocator_set_it_branch_type (GumThumbRelocator * self,
+                                        GumThumbRelocatorBranchType type)
 {
-  return (insn >> 4) & 0xf;
-}
+  switch (type)
+  {
+    case GUM_THUMBRELOCATOR_BRANCH_SHORT:
+    case GUM_THUMBRELOCATOR_BRANCH_LONG:
+      self->it_branch_type = type;
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:985]: No need for the `switch` here, we shouldn't be catching programmer errors.

File: gum/arch-arm/gumthumbrelocator.c [hunk 5]
```diff
@@ -969,8 +974,19 @@ gum_parse_it_instruction_block_size (guint16 insn)
...
+      break;
+    default:
+      g_assert_not_reached ();
+      break;
+  }
+
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.c:991]:
```suggestion
```

File: gum/arch-arm/gumthumbrelocator.h [hunk 6]
```diff
@@ -16,6 +16,14 @@ G_BEGIN_DECLS
 typedef struct _GumThumbRelocator GumThumbRelocator;
 typedef struct _GumThumbLocation GumThumbLocation;
 typedef struct _GumThumbITBlock GumThumbITBlock;
+typedef guint GumThumbRelocatorBranchType;
+
+enum _GumThumbRelocatorBranchType
+{
+  GUM_THUMBRELOCATOR_BRANCH_INVALID,
+  GUM_THUMBRELOCATOR_BRANCH_SHORT,
+  GUM_THUMBRELOCATOR_BRANCH_LONG
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.h:25]: How about `GumITBranchType`:

```suggestion
  GUM_IT_BRANCH_INVALID,
  GUM_IT_BRANCH_SHORT,
  GUM_IT_BRANCH_LONG
```

File: gum/arch-arm/gumthumbrelocator.h [hunk 7]
```diff
@@ -16,6 +16,14 @@ G_BEGIN_DECLS
...
+};
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.h:26]: This should also be moved last so the order is consistent with the typedefs.

File: gum/arch-arm/gumthumbrelocator.h [hunk 8]
```diff
@@ -92,6 +101,8 @@ GUM_API gboolean gum_thumb_relocator_can_relocate (gpointer address,
     guint min_bytes, GumRelocationScenario scenario, guint * maximum);
 GUM_API guint gum_thumb_relocator_relocate (gpointer from, guint min_bytes,
     gpointer to);
+void gum_thumb_relocator_set_it_branch_type (GumThumbRelocator * self,
+    GumThumbRelocatorBranchType type);
```
Comment (oleavr) [inline gum/arch-arm/gumthumbrelocator.h:105]: Should moved right after `reset()`.

File: gum/arch-arm/gumthumbwriter.h [hunk 9]
```diff
@@ -54,6 +54,8 @@ GUM_API gboolean gum_thumb_writer_flush (GumThumbWriter * self);
 
 GUM_API gboolean gum_thumb_writer_put_label (GumThumbWriter * self,
     gconstpointer id);
+GUM_API gboolean gum_thumb_writer_commit_label_by_id (GumThumbWriter * self,
```
Comment (oleavr) [inline gum/arch-arm/gumthumbwriter.h:57]:
```suggestion
GUM_API gboolean gum_thumb_writer_commit_label (GumThumbWriter * self,
```

File: tests/core/arch-arm/stalker-arm.c [hunk 10]
```diff
@@ -1649,6 +1650,41 @@ TESTCASE (thumb_it_eq_pop)
   GUM_ASSERT_EVENT_ADDR (ret, 2, location, func + 14);
 }
 
+TESTCODE (thumb_itttt_eq_blx_reg,
+  0x00, 0xb5,             /* push {lr}          */
+  0x00, 0x1a,             /* subs r0, r0, r0    */
```
Comment (oleavr) [inline tests/core/arch-arm/stalker-arm.c:1655]:
```suggestion
  0x00, 0x1a, /* subs r0, r0, r0 */
```

For consistency.

File: tests/core/arch-arm/stalker-arm.c [hunk 11]
```diff
@@ -1649,6 +1650,41 @@ TESTCASE (thumb_it_eq_pop)
...
+  /* part_two:                                  */
+  0x00, 0x28,             /* cmp r0, #0         */
+  0x01, 0xbf,             /* itttt eq           */
+  0x02, 0x30,             /* adds r0, #2        */
+  0x02, 0x30,             /* adds r0, #2        */
+  0x02, 0x30,             /* adds r0, #2        */
+  0x88, 0x47,             /* blx r1             */
+  0x00, 0xbd,             /* pop {pc}           */
+
+
```
Comment (oleavr) [inline tests/core/arch-arm/stalker-arm.c:1675]:
```suggestion
```

**Comment by WorksButNotTested** [comment]
Should be ready for another look.

**Pull request merged**